### PR TITLE
legion: add mpi interoperability and shared libs

### PR DESCRIPTION
**Pull request by junghans**

### CODE REVIEW

File: var/spack/repos/builtin/packages/legion/package.py [hunk 1]
```diff
@@ -47,9 +47,12 @@ class Legion(CMakePackage):
     version('17.02.0', '31ac3004e2fb0996764362d2b6f6844a')
 
     variant('debug', default=False, description='Build debug version')
+    variant('mpi', default=False,
+            description='Build on top of mpi conduit for mpi inoperability')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/legion/package.py]: My only concern here is whether or not `mpi` should default to True. There's a lot of bugs in the concretizer right now that make it hard for package A to default to `+mpi` and its dependency package B to default to `~mpi`. In #2590 I made sure that every `mpi` variant defaulted to True to workaround this until the concretizer is fixed. Would you be opposed to having the default be True?

Comment (junghans) [inline var/spack/repos/builtin/packages/legion/package.py]: Sure!

**Comment by junghans** [comment]
@adamjstewart: What is your take on `BUILD_SHARED_LIBS`? Default on or off?

(Background: gasnet is static only, even without `-fPIC`(I will make a PR that), and hence legion has static libs as default, but flecsi prefers a shared version of legion to not have to worry about legion dependency libraries.)

**Comment by adamjstewart** [comment]
@junghans I usually build both static and shared if possible, otherwise I build just shared. A lot of other users will add variants to control this. I personally don't have any strong feelings on this, but @davydden and @citibeth do.

**Comment by junghans** [comment]
@citibeth : Please review the shared variant part.

**Pull request merged**